docs/devel/infrastructure-interaction.md

```diff
@@ -0,0 +1,85 @@
+
+1. Infrastructure admin.  The person with the permission and ability to make changes to the infrastructure.
+2. Kube admin.  The person with the permission to change how the kube cluster itself operates.
+3. Kube user.  The person who just wants to write and run an app.
+
+My actual design decoupling can be broken down into 3 steps.
+
+1. The “infrastructure admin”, needs to make these external functions available to kube, potentially via a cloud provider interface similar to how minions are handled today.  We would want something similar for storage and external reachability functionality.  It seems likely that we may want to decompose the cloud_provider into multiple smaller providers.  ‘minion_provider’, ‘storage_provider’, etc.  But that is a specific detail to work out per infrastructure concept.
+
+2. The “kube admin” needs to explicitly tell the cluster to use the functionality exposed by the infrastructure. In the case of minions this defaults to “*” but we have the –minion_regexp. That means minions sorta fit this model. GCE PDs fit as you have to specify the GCEPersistentDisk.  The specifics of how the kube admin tells the cluster to use the functionality will likely be functionality specific.  (Although I’ll give my thoughts on some of these things below)
+
+3. “Something” needs to watch the cluster to see how the resource is supposed to be used and make that happen. Minions fall down here, it's more that the apiserver reaches out to them, but I think that’s an ok exception.  The model still mostly fits for GCE PDs. The kubelet watches the pod description and from the volume info in the pod definition it mounts and actually uses the storage. I think a different, but very similar, pattern should be generalized and should work well for externally reachable functionality (load balancers, external IP, etc) and other storage mechanisms.
+
+I'll work through a couple of examples of what I'm thinking. Say an external load balancer (conceptually should map to external IPs or whatever else) and an NFS server (should map to other storage, including GCE PDs)
+
+**External Load Balancer Example**
+
+If I install an external load balancer into my datacenter with some form of reachability into the cluster (aka it can reach pods via pod IP addresses) I believe that the cloud provider interface needs to tell the cluster about the load balancer via some sort of “name” or “label” or whatever detail we agree upon. That's step 1. Step 2 is, in my mind, creating a new resource in kube. This resource should be a mapping between a service/pod/label and the 'name'/'label' of the external device. Yes, I’m talking about a first class resource like a pod, service, minion, etc. Step 3 is that the external load balancer (or a plugin in front of the load balancer) should watch the api server for new assignments and make the magic configuration changes for my load balancer.
```

Comment:
Agree we'll need a real resource. It should have the same metadata as our other objects (names, labels, annotations, etc.).